outputs/awssecuritylake.go

```diff
@@ -196,54 +198,80 @@ func getAWSSecurityLakeSeverity(priority types.PriorityType) (int32, string) {
+
 	for {
-		time.Sleep(time.Duration(c.Config.AWS.SecurityLake.Interval) * time.Minute)
-		// time.Sleep(5 * time.Second)
-		batch := make([]memlog.Record, c.Config.AWS.SecurityLake.BatchSize)
-		count, err := c.Config.AWS.SecurityLake.Memlog.ReadBatch(c.Config.AWS.SecurityLake.Ctx, *c.Config.AWS.SecurityLake.ReadOffset+1, batch)
+		time.Sleep(time.Duration(awslake.Interval) * time.Minute)
+		batch := make([]memlog.Record, awslake.BatchSize)
+		count, err := ml.ReadBatch(ctx, *awslake.ReadOffset+1, batch)
 		if err != nil {
-			if err.Error() != "future offset" {
+			if errors.Is(err, memlog.ErrFutureOffset) {
+				// 	could be partial batch, continue batch logic below
+			}
+
+			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
 				go c.CountMetric(Outputs, 1, []string{"output:awssecuritylake.", "status:error"})
 				c.Stats.AWSSecurityLake.Add(Error, 1)
 				c.PromStats.Outputs.With(map[string]string{"destination": "awssecuritylake.", "status": Error}).Inc()
 				log.Printf("[ERROR] : %v SecurityLake. - %v\n", c.OutputType, err)
+				return
+			}
+
+			if errors.Is(err, memlog.ErrOutOfRange) {
+				earliest, _ := ml.Range(ctx)
+
+				go c.CountMetric(Outputs, 1, []string{"output:awssecuritylake.", "status:error"})
+				c.Stats.AWSSecurityLake.Add(Error, 1)
+				c.PromStats.Outputs.With(map[string]string{"destination": "awssecuritylake.", "status": Error}).Inc()
+
+				msg := fmt.Errorf("slow batch reader: resetting read offset from %d to %d: %v",
+					*awslake.ReadOffset,
+					earliest,
+					err,
+				)
+				log.Printf("[ERROR] : %v SecurityLake. - %v\n", c.OutputType, msg)
+
+				awslake.ReadOffset = &earliest
 				continue
 			}
+
+			// catch all and return
+			go c.CountMetric(Outputs, 1, []string{"output:awssecuritylake.", "status:error"})
+			c.Stats.AWSSecurityLake.Add(Error, 1)
+			c.PromStats.Outputs.With(map[string]string{"destination": "awssecuritylake.", "status": Error}).Inc()
+			log.Printf("[ERROR] : %v SecurityLake. - %v\n", c.OutputType, err)
+			return
 		}
+
 		if count > 0 {
-			var err error
-			*c.Config.AWS.SecurityLake.ReadOffset = batch[count-1].Metadata.Offset
 			uid := uuid.New().String()
-			if count == 1 {
-				err = c.writeParquet(uid, []memlog.Record{batch[0]})
-			}
-			if count > 1 {
-				err = c.writeParquet(uid, batch[:count-1])
-			}
-			if err != nil {
+
+			if err := c.writeParquet(uid, batch[:count]); err != nil {
 				go c.CountMetric(Outputs, 1, []string{"output:awssecuritylake.", "status:error"})
 				c.Stats.AWSSecurityLake.Add(Error, 1)
 				c.PromStats.Outputs.With(map[string]string{"destination": "awssecuritylake.", "status": Error}).Inc()
-				continue
-			}
-			if err != nil {
-				go c.CountMetric(Outputs, 1, []string{"output:awssecuritylake.", "status:ok"})
-				c.Stats.AWSSecurityLake.Add(OK, 1)
-				c.PromStats.Outputs.With(map[string]string{"destination": "awssecuritylake.", "status": "ok"}).Inc()
+				return // treat as non-recoverable here
```

Comment:
Please let me know how to handle Parquet write errors. To me feels unrecoverable, not sure how main handles this though.

Reply:
we pass to next bath, that's all